### PR TITLE
Synchronized training (EASGD)

### PR DESCRIPTION
**Pull request by rizar**
Contains my work in progress on an extension for Platoon. Includes #37, will be rebased upon the merge of that PR.

### CODE REVIEW

**Comment by rizar** [comment]
[Usage example:](https://github.com/rizar/blocks-examples/blob/use_synchronize/mnist/__init__.py). Disclaimer: can not get any benefit from EAGS so far :(

My reasoning is the following: even though communication is definitely overwhelming for the simple MNIST demo that I use, I should see much better results after running two training processes for N epochs in parallel than after running one process for the same number of epochs. So far, it only gets worse when I use more than one process.

**Comment by rizar** [comment]
@nouiz @abergeron  @carriepl : you can track the progress here.

**Comment by carriepl** [comment]
@rizar : What {alpha, number of processes, sync_freq} combinations did you try?

**Comment by carriepl** [comment]
Also, FYI, there is a PR under review to rename Soldier and Lieutenant to more neutral names. Obviously, it would have an effect on this PR. https://github.com/mila-udem/platoon/pull/22

**Comment by rizar** [comment]
Sure, thanks for the heads-up, I will rename everything as soon as you guys
merge the PR.

**Comment by carriepl** [comment]
@rizar : the PR has been merged

**Comment by rizar** [comment]
@sotelo, this is WIP for data-parallel training

**Comment by rizar** [comment]
I ran it to convergence, and we can see that the workers do not help each other at all. Blue line is the progress of a job working alone, red and green are two workers working in parallel. Test error is displayed.

plot

**Comment by rizar** [comment]
This was rebased and can now be used with the latest master.

**Comment by nouiz** [comment]
https://github.com/abergeron/platoon/commit/afcea0ff2014156ff9d8377f48f6ba1542fe1750 when the corresponding PR is merged, we can have the worker close correction the connection. It would be great to update this PR with that.

File: blocks_extras/extensions/synchronization.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                self.job_name, self.main_loop.model.parameters,
+                self.sync_rule, cleanup=self.worker.is_main_worker)
+            if not self.worker.is_main_worker:
+                self.worker.copy_to_local()
+                logger.debug("Copied parameters from shared")
+            else:
+                logger.debug("Initialized shared parameters")
+        elif (which_callback == 'after_batch' or
+              which_callback == 'after_epoch'):
+            self.worker.sync_params(synchronous=True)
```
Comment (nouiz) [inline blocks_extras/extensions/synchronization.py]: maybe this could be a parameter.

File: blocks_extras/extensions/synchronization.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    @property
+    def is_main_worker(self):
+        if not hasattr(self, '_is_main_worker'):
+            self._is_main_worker = self.send_req('is_main_worker?')
+        return self._is_main_worker
+
+    @property
+    def seed(self):
+        if not hasattr(self, '_is_seed'):
```
Comment (nouiz) [inline blocks_extras/extensions/synchronization.py]: bug here

**Comment by rizar** [comment]
For the record, with ASGD I do get the 2x speedup  on 2 GPUs.

File: blocks_extras/extensions/synchronization.py (hunk 1 above)
Comment (rizar) [inline blocks_extras/extensions/synchronization.py]: Maybe later, for now this seems like a dangerous option.

**Comment by rizar** [comment]
Status update: this PR is ready to be reviewed, except for `blocks-parallel`. In fact I think that we should not merge blocks-parallel, because the users can quickly implement scripts like these using their scripting language of choice. 

@bartvm , would you like to take a look?

File: blocks_extras/extensions/synchronization.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+to share parameters and perform communication among multiple training
+jobs. Below is the list of the key concepts introduced in Platoon:
+
+- the training process is called *worker*
+- a separate process called *controller* is sending/receiving messages
+  from/to workers and is responsible for synchronizing their activities
+- each process has its *local parameters*, which are usual Theano shared
+  variables
+- all processes share *global* parameters, which are stored in
+  shared memory (not to be confused with shared Theano variables!)
```
Comment (nouiz) [inline blocks_extras/extensions/synchronization.py]: I would specify CPU shared memory. The GPU have a shared memory, but it is a different concept.

**Comment by nouiz** [comment]
This can be reviewed, but there is a PR to platoon that change a little the
interface. So before merging this, wait for this PR to be merged and this
one to be updated:

https://github.com/mila-udem/platoon/pull/29

**Comment by rizar** [comment]
OK, no problem, I can wait.

**Comment by nouiz** [comment]
The other PR is now merged.

**Comment by rizar** [comment]
@abergeron , this PR is functional again, you can review it.

File: blocks_extras/extensions/synchronization.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    It is recommended to trigger an additional
+    synchronization after every time consuming operation, such as
+    validation or checkpointing. Otherwise you might end up using
+    very stale parameters.
+
+    """
+    def __init__(self, worker, **kwargs):
+        kwargs.setdefault("before_training", True)
+        kwargs.setdefault("on_resumption", True)
```
Comment (abergeron) [inline blocks_extras/extensions/synchronization.py]: Shouldn't you do something with on_resumption in do()?

Comment (rizar) [inline blocks_extras/extensions/synchronization.py]: Fixed below, but not tested.

**Pull request merged**